### ISSUE TITLE
Return std::result::Result instead of panic!()ing

### ISSUE THREAD

**Issue by ckaran** [SergiusIW/collider-rs#10]
I've noticed that there are a number of places where code can `panic!()` instead of returning `std::result::Result`.  Would it be a lot of effort to convert all of these locations to return `std::result::Result` instead?  I'm trying to create python bindings to Collider, and the panics/aborts force me to write checks into the wrapper code that will be repeated within Collider anyways.  A good tool to help find paths that lead to panics is [rustig](https://github.com/Technolution/rustig).

**Comment by SergiusIW** [comment]
The asserts and panics in this codebase seem to fall into a few categories:
1. In unit tests
2. Internal sanity checks that should never fail
3. Improper usage of the library (things like illegal arguments, removing a hitbox that isn't there, encountering very extreme floating point numbers)

This is more or less in line with Rust's guidelines for when to use panic. See https://doc.rust-lang.org/book/ch09-03-to-panic-or-not-to-panic.html

> It’s advisable to have your code panic when it’s possible that your code could end up in a bad state. In this context, a bad state is when some assumption, guarantee, contract, or invariant has been broken, such as when invalid values, contradictory values, or missing values are passed to your code—plus one or more of the following:
> 
> * The bad state is not something that’s expected to happen occasionally.
> * Your code after this point needs to rely on not being in this bad state.
> * There’s not a good way to encode this information in the types you use.
> 
> If someone calls your code and passes in values that don’t make sense, the best choice might be to call panic! and alert the person using your library to the bug in their code so they can fix it during development.

Did you have any specific instances of panicking in the codebase that you think would be better expressed as a Result?

**Comment by ckaran** [comment]
> Did you have any specific instances of panicking in the codebase that you think would be better expressed as a Result?

I think so.  Here is the list of the ones I've found so far.

- In `Collider`
  - `set_time()` - Panics if the incoming time is outside of the range [`self.time()`, `self.next_time()`].  If you're running multiple Collider instances in parallel, advancing them all to the same point in time at the same time, then this could be problematic.
  - `add_hitbox()` - Panics if you reuse an id.
  - All of the following panic if you pass in an id that the Collider instance doesn't know about.
    - `get_hitbox()`
    - `set_hitbox_vel()`
    - `remove_hitbox()`
    - `get_overlaps()`
    - `is_overlapping()`
    - `query_overlaps()`
- In `Shape`
  - `new()` - Panics (via an `assert!()`) if the dimensions are negative
  - `with_any_dims()` - Panics (via an `assert!()`) if the shape is a circle and the dimensions are not equal.

I'm sure that there are others, but that is why I think that `rustig` would be a good tool to use, so that you can find all the paths that can panic.

**Comment by SergiusIW** [comment]
I would argue that all of these cases signal that the user is misusing the library, indicating bugs in the user's code. If there is a bug in the user's code, I think panicking is more appropriate than returning a Result or Option. I believe this to be in line with Rust's best practices, as I have stated above.

**Comment by ckaran** [comment]
I agree with you that those are examples of misuse of the library.  The issue is how to recover from the misuse.  When Collider `panic!()`s, it also kills the python interpreter, which is difficult or impossible for a python coder to recover from.  My current way around the problem is to create checks in my wrapper code that duplicate the checks found in Collider, and to raise exceptions that way.  I see this as inefficient, but I do see your point about how in pure rust code, there should be a `panic!()`...

OK, would you be open to modifying Collider by splitting up the validation and execution portions of functions into two parts?  I'm thinking of something like the following (please excuse any typos):

```rust

pub fn validate_foo(...) -> Result<()> {
    // All the validation code bits return Result
    Ok(())
}

pub fn foo(...) -> ... {
    match validate_foo(...) {
        Ok(_) => {
            // Whatever foo is supposed to do in here;
        },
        Err(_) => panic!()
    }
}
```

The validation code is then accessible from python and any other wrappers, which means that if new validation code is added, the wrapper gets it automatically.  The rust code can still choose to `panic!()` if there are any problems.  This still leaves the problem of the validation code being run twice from python, but I think that is something end users will just need to deal with.
